test-control/Test/Control/RefCount.hs

```diff
@@ -79,3 +96,79 @@ prop_refCount = once $ ioProperty $ do
 #else
     check = \case Left (AssertionFailed _) -> False; Right () -> True
 #endif
+
+#ifdef NO_IGNORE_ASSERTS
+data TestObject = TestObject !(RefCounter IO)
+
+instance RefCounted TestObject where
+    type FinaliserM TestObject = IO
+    getRefCounter (TestObject rc) = rc
+
+data TestObject2 = TestObject2 (Ref TestObject)
+
+instance RefCounted TestObject2 where
+    type FinaliserM TestObject2 = IO
+    getRefCounter (TestObject2 (DeRef to1)) = getRefCounter to1
+
+prop_ref_double_free :: Property
+prop_ref_double_free = once $ ioProperty $ do
+    finalised <- newIORef False
+    ref <- newRef (writeIORef finalised True) TestObject
+    releaseRef ref
+    True <- readIORef finalised
+    Left RefDoubleRelease{} <- try $ releaseRef ref
+    checkForgottenRefs
+
+prop_ref_use_after_free :: Property
+prop_ref_use_after_free = once $ ioProperty $ do
+    finalised <- newIORef False
+    ref <- newRef (writeIORef finalised True) TestObject
+    releaseRef ref
+    True <- readIORef finalised
+    Left RefUseAfterRelease{} <- try $ withRef ref return
+    Left RefUseAfterRelease{} <- try $ case ref of DeRef x -> evaluate x
+    Left RefUseAfterRelease{} <- try $ dupRef ref
+    checkForgottenRefs
+
+prop_ref_never_released0 :: Property
+prop_ref_never_released0 = once $ ioProperty $ do
+    finalised <- newIORef False
+    ref <- newRef (writeIORef finalised True) TestObject
+    _ <- case ref of DeRef x -> evaluate x
+    checkForgottenRefs
+    -- ref is still being used, so check should not fail
+    _ <- case ref of DeRef x -> evaluate x
+    releaseRef ref
+
+prop_ref_never_released1 :: Property
+prop_ref_never_released1 =
+    once $ ioProperty $
+    handle expectRefNeverReleased $ do
+      finalised <- newIORef False
+      ref <- newRef (writeIORef finalised True) TestObject
+      _ <- withRef ref return
+      _ <- case ref of DeRef x -> evaluate x
+      -- ref is never released, so should fail
+      checkForgottenRefs
```

Comment:
No need because we use
```
  handle expectRefNeverReleased $ do
```
around the whole test.

This is more robust, because the `RefNeverReleased` can be thrown aywhere, by any `Ref` operation. It is not just thrown by `checkForgottenRefs`.

This is because the new impl relies on GC finalisers that update a global var, and then all the other `Ref` operations poll that global var for more prompt failure detection.

`checkForgottenRefs` is more likely to trigger the exception because it runs GC and then polls the global var.